### PR TITLE
Remove some D2D memcpy in T5 BeamSearch

### PR DESCRIPTION
**Pull request by wangyems**
### Description

replace D2D memcpy with cuda kernels in creating decoder initial feeds
before:
image

after:
image

### CODE REVIEW

**Review by zhanghuanrong — DISMISSED** [review]

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 1]
```diff
@@ -805,6 +806,189 @@ void UpdateDecoderMaskedMultiHeadAttentionCacheIndirection(int32_t* tgt_indir_ca
                                                                                           max_seq_length,
                                                                                           current_length);
 }
+
+#ifndef USE_ROCM
+namespace {
+template <typename T, size_t size>
+struct TypeMapper : public V_vec_m_<T, size> {};
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:813]: >V_vec_m_ [](http://example.com/codeflow?start=27&length=8)

It is better to expand the V_vec_m_ directly

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 2]
```diff
@@ -805,6 +806,189 @@ void UpdateDecoderMaskedMultiHeadAttentionCacheIndirection(int32_t* tgt_indir_ca
...
+void KeyCacheExpansionKernelLauncher(const T* key_cache,
+                                     T* key_cache_expanded,
+                                     int batch_size,
+                                     int beam_width,
+                                     int num_heads,
+                                     int sequence_length,
+                                     int max_seq_length,
+                                     int head_size,
+                                     cudaStream_t stream) {
+  const dim3 block(256);
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:860]: >block(256); [](http://example.com/codeflow?start=13&length=11)

block size is too large. head_size is usually 32 or 64.

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 1 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:813]: currently the int mapper only applies to this case

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 2 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:860]: we can use nearest 2^^n

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 3]
```diff
@@ -805,6 +806,203 @@ void UpdateDecoderMaskedMultiHeadAttentionCacheIndirection(int32_t* tgt_indir_ca
...
+  int tpb = std::max(32, equiv_head_size);
+
+  // round up tpb to power of 2
+  --tpb;
+  tpb |= (tpb >> 1);
+  tpb |= (tpb >> 2);
+  tpb |= (tpb >> 4);
+  tpb |= (tpb >> 8);
+  tpb |= (tpb >> 16);
+  tpb++;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:875]: you need to set block to this or use tpb in kernel launching directly.

Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:875]: you can check the last commit

**Review by yufenglee — APPROVED** [review]

**Pull request merged**